sql/data_source.go

```diff
@@ -0,0 +1,497 @@
+//
+// A data source is an object that can deliver rows of column data,
+// implemented in CockroachDB as an array of values.
+// The defining property in a data source is that the columns in its
+// result row arrays are always numbered from 0.
+//
+// From the language perspective, data sources are defined indirectly by:
+// - the FROM clause in a SELECT statement
+// - JOIN clauses within the FROM clause
+// - the clause that follows INSERT INTO colName(Cols...)
+// - the clause that follows UPSERT ....
+// - the invisible data source defined by the original table row during
+//   UPSERT, if it exists.
+//
+// Most expressions (parser.Expr trees) in CockroachDB refer to a
+// single data source. A notable exception is UPSERT, where expressions
+// can refer to two sources: one for the values being inserted, one for
+// the original row data in the table for conflicting (already
+// existing) rows.
+//
+// Meanwhile, qvalues in CockroachDB provide the interface between
+// symbolic names in expressions (e.g. "f.x", called QualifiedNames,
+// or qnames) and data sources. During evaluation, a qvalue must
+// resolve to a column value. For a given qname there are thus two
+// subsequent questions that must be answered:
+//
+// - which data source is the qname referring to? (when there is more than 1 source)
+// - which 0-indexed column in that data source is the qname referring to?
+//
+// The qvalue must distinguish data sources because the same column index
+// may refer to different columns in different data sources. For
+// example in an UPSERT statement the qvalue for "excluded.x" could refer
+// to column 0 in the (already existing) table row, whereas "src.x" could
+// refer to column 0 in the valueNode that provides values to insert.
+//
+// Within this context, the infrastructure for data sources and qvalues
+// is implemented as follows:
+//
+// - dataSourceInfo provides column metadata for exactly one data source
+// - the columnRef in qvalues contains a link (pointer) to the
+//   dataSourceInfo for its data source
+// - qvalResolver (select_qvalue.go) is tasked with linking back qvalues with
+//   their data source and column index.
+//
+// This being said, there is a misunderstanding one should be careful
+// to avoid: *there is no direct relationship between data sources and
+// table names* in SQL. In other words:
+//
+// - the same table name can be present in two or more data sources; for example
+//   with:
+//        INSERT INTO excluded VALUES (42) ON CONFLICT (x) DO UPDATE ...
+//   the name "excluded" can refer either to the data source for VALUES(42)
+//   or the implicit data source corresponding to the rows in the original table
+//   that conflict with the new values.
+//
+//   When this happens, a qname of the form "excluded.x" must be resolved
+//   by considering all the data sources; if there is more than one data
+//   source providing the table name "excluded" (as in this case), an ambiguity
+//   is reported.
```

Comment:
An alternative to reporting an ambiguity is to precisely define the resolution process, similar to the way a variable is resolved from inner scopes to outer scopes. I mention this because that's my understanding of what Postgres does and we likely want to be compatible in this regard. The resolution process might also be specified by the SQL standard.

Note that I'm referring to the name resolution process in general, not just the specific example of `excluded` within an `ON CONFLICT` expression.